sdk/include/ndk/haltypes.h

```diff
@@ -181,22 +1729,336 @@ typedef struct _HAL_PRIVATE_DISPATCH
+    pKdEnumerateDebuggingDevices KdEnumerateDebuggingDevices;
+    pHalFlushIoRectangleExternalCache HalFlushIoRectangleExternalCache;
+    pHalPowerEarlyRestore HalPowerEarlyRestore;
+    pHalQueryCapsuleCapabilities HalQueryCapsuleCapabilities;
+    pHalUpdateCapsule HalUpdateCapsule;
+    pHalPciMultiStageResumeCapable HalPciMultiStageResumeCapable;
+    pHalDmaFreeCrashDumpRegisters HalDmaFreeCrashDumpRegisters;
+    pHalAcpiAoacCapable HalAcpiAoacCapable;
+#endif
+#if (NTDDI_VERSION >= NTDDI_WINBLUE)
+    pHalInterruptSetDestination HalInterruptSetDestination;
+    pHalGetClockConfiguration HalGetClockConfiguration;
+    pHalClockTimerActivate HalClockTimerActivate;
+    pHalClockTimerInitialize HalClockTimerInitialize;
+    pHalClockTimerStop HalClockTimerStop;
+    pHalClockTimerArm HalClockTimerArm;
+    pHalTimerOnlyClockInterruptPending HalTimerOnlyClockInterruptPending;
+    pHalAcpiGetMultiNode HalAcpiGetMultiNode;
+    pHalPowerSetRebootHandler HalPowerSetRebootHandler;
+    pHalIommuRegisterDispatchTable HalIommuRegisterDispatchTable;
+    pHalTimerWatchdogStart HalTimerWatchdogStart;
+    pHalTimerWatchdogResetCountdown HalTimerWatchdogResetCountdown;
+    pHalTimerWatchdogStop HalTimerWatchdogStop;
+    pHalTimerWatchdogGeneratedLastReset HalTimerWatchdogGeneratedLastReset;
+    pHalTimerWatchdogTriggerSystemReset HalTimerWatchdogTriggerSystemReset;
+#endif
+#if (NTDDI_VERSION >= NTDDI_WIN10)
+    pHalInterruptVectorDataToGsiv HalInterruptVectorDataToGsiv;
+    pHalInterruptGetHighestPriorityInterrupt HalInterruptGetHighestPriorityInterrupt;
+    pHalProcessorOn HalProcessorOn;
+    pHalProcessorOff HalProcessorOff;
+    pHalProcessorFreeze HalProcessorFreeze;
+    pHalDmaLinkDeviceObjectByToken HalDmaLinkDeviceObjectByToken;
+    pHalDmaCheckAdapterToken HalDmaCheckAdapterToken;
+    pHalPciLateRestore HalPciLateRestore;
+    pHalTimerConvertPerformanceCounterToAuxiliaryCounter HalTimerConvertPerformanceCounterToAuxiliaryCounter;
+    pHalTimerConvertAuxiliaryCounterToPerformanceCounter HalTimerConvertAuxiliaryCounterToPerformanceCounter;
+    pHalTimerQueryAuxiliaryCounterFrequency HalTimerQueryAuxiliaryCounterFrequency;
+    pHalConnectThermalInterrupt HalConnectThermalInterrupt;
+    pHalIsEFIRuntimeActive HalIsEFIRuntimeActive;
 #endif
 } HAL_PRIVATE_DISPATCH, *PHAL_PRIVATE_DISPATCH;
 
+typedef struct _IOMMU_DEVICE_PATH
+{
+    GUID BusTypeGuid;
+    ULONG UniqueIdLength;
+    ULONG PathLength;
+    PVOID UniqueId;
+    PVOID Path;
+} IOMMU_DEVICE_PATH, *PIOMMU_DEVICE_PATH;
+
+typedef union _IOMMU_SVM_CAPABILITIES
+{
+    struct {
+        ULONG AtsCapability:1;
+        ULONG PriCapability:1;
+        ULONG PasidCapability:1;
+        struct {
+            ULONG PasidMaxWidth:5;
+            ULONG PasidExePerm:1;
+            ULONG PasidPrivMode:1;
+            ULONG AtsPageAlignedRequest:1;
+            ULONG AtsGlobalInvalidate:1;
+            ULONG AtsInvalidateQueueDepth:5;
```

Comment:
Please use a uniform style for the bitfields in this file, like:
`AtsInvalidateQueueDepth : 5;` , or: `AtsInvalidateQueueDepth: 5;` , which makes them more easily readable.
(This was already done elsewhere for very few structs.)